### ISSUE TITLE
How can I share the weights between two different dilations cnn layer in tensorflow2.0

### ISSUE THREAD

**Issue by HuaYZhao** [tensorflow/tensorflow#37278]
How can I share the weights between two different dilations cnn layer in tensorflow2.0
In tensorflow1.x, I can just use the tf.variable_scope with the tf.AUTO_REUSE

**Comment by pandrey-fr** [comment]
Hi,
One of the easiest ways to share weights is to actually call the same `Layer` instance twice, _i.e._ running `cnn_layer(cnn_layer(inputs))` is strictly similar to using two similarly-specified layers with shared weights.
Of course, this may be more or less easy to do depending on your code and context. If you want better advice, please share some details and example code - although to be honest this would probably be better asked on StackOverflow as this is more a use-case question than an issue _per se_ :)

**Comment by HuaYZhao** [comment]
Hello, thank you for your reply!
```
atro_cnn1 = Conv1D(dilation_rate=1,...)
atro_cnn2 = Conv1D(dilation_rate=3,...)

x = atro_cnn1(x)
x = atro_cnn2(x)
```

`atro_cnn1`  and `atro_cnn2`  are the same except for `dilation_rate`. In dilated convolution, `dilation_rate` does not affect the parameters of the layer. So I want `atro_cnn1`  and `atro_cnn2` to be able to use the same parameter. But in tf2, it seems that I can't get past reuse and change the `dilation_rate` of that object.

**Comment by pandrey-fr** [comment]
Thank you for clarifying!
There probably could be a simpler way, but I guess one way is to implement a custom `Layer` that takes a parent one and re-uses its weights while allowing to change some of the convolution operation's parameters. Here is a home-made implementation of the sort (I did not fully test it, let me know if it works). I tried to generalize it to both Conv1D, Conv2D and Conv3d; the idea is that any of the `__init__` parameter set to None re-uses the value of the parent layer (in your case, `atro_cnn1`).
```python
class SharedWeightsConv(tf.keras.layers.Layer):
    
    def __init__(
            self,
            parent,
            strides=None,
            padding=None,
            dilation_rate=None,
            activation=None,
            **kwargs
        ):
        conv_classes = (
            tf.keras.layers.Conv1D,
            tf.keras.layers.Conv2D,
            tf.keras.layers.Conv3D
        )
        if not any(isinstance(parent, cls) for cls in conv_classes):
            raise TypeError("'parent' should be a keras convolution layer.")
        super().__init__(**kwargs)
        self.parent = parent
        self.rank = parent.rank
        self.activation = (
            parent.activation if activation is None 
            else tf.keras.activations.get(activation)
        )
        cnn_kwargs = {
            'strides': strides,
            'padding': padding,
            'data_format': None,
            'diltion_rate': dilation_rate,
        }
        self.cnn_kwargs = {
            key: getattr(parent, key) if value is None else value
            for key, value in cnn_kwargs
        }
        self.built = self.parent.built
        self.cnn_op = {
            1: tf.keras.backend.conv1d,
            2: tf.keras.backend.conv2d,
            3: tf.keras.backend.conv3d
        }.get(self.rank)

    def build(self, input_shape):
        if not self.built:
            self.parent.build(input_shape)
        self.built = True

    def call(self, inputs):  # adapted from Conv parent layer
        if self.cnn_kwargs['padding'] == 'causal' and self.rank == 1:
            inputs = tf.pad(inputs, self._compute_causal_padding())
        outputs = self.cnn_op(inputs, self.parent.kernel, **self.cnn_kwargs)
        if self.parent.use_bias:
            if self.cnn_kwargs['data_format'] == 'channels_first':
                if self.rank == 1:
                    shape = (1, self.parent.filters, 1)
                    outputs += tf.reshape(self.parent.bias, shape)
                else:
                    outputs = tf.nn.bias_add(
                        outputs, self.parent.bias, data_format='NCHW'
                    )
            else:
                outputs = tf.nn.bias_add(
                    outputs, self.parent.bias, data_format='NHWC'
                )
        if self.activation is not None:
            return self.activation(outputs)
        return outputs

    def _compute_causal_padding(self):  # adapted from Conv parent layer
        left_pad = self.cnn_kwargs['dilation_rate'][0]
        left_pad *= (self.parent.kernel_size[0] - 1)
        if self.cnn_kwargs['data_format'] == 'channels_last':
          causal_padding = [[0, 0], [left_pad, 0], [0, 0]]
        else:
          causal_padding = [[0, 0], [0, 0], [left_pad, 0]]
        return causal_padding
```

For your example:
```python
atro_cnn1 = Conv1D(dilation_rate=1, ...)
atro_cnn2 = SharedWeightsConv(atro_cnn1, dilation_rate=1)

x = atro_cnn1(x)
x = atro_cnn2(x)
```

I hope this helps, let me know if bugs arise!

**Comment by HuaYZhao** [comment]
Thank you very much for your answer. I think I've got your point, and it should be a workable approach, although it seems a bit cumbersome. 
I will try to really test the feasibility of this method in my later experiments, in my scenario. 
For some more complex layers (customizations), such a shared approach does not seem to be easy to implement. 
As you might expect, I also think there should be some more intuitive approach here (or will be implemented).

**Comment by pandrey-fr** [comment]
I totally agree; I had to write some similarly cumbersome custom layers for some personal projects, and I really hope a better solution could be exposed or implemented by a tensorflower soon...

**Comment by Susmit-A** [comment]
A simpler way to share weights using a custom layer would be to use `tf.nn.conv2d`:

class SharedDilatedConv(tf.keras.layers.Layer):
     def __init__(
            self,
            filters,
            kernel_size,
            strides=None,
            padding=None,
            dilation_rates=None,
            activation=None,
            use_bias=True,
            **kwargs
        ):
        self.filters = filters
        self.kernel_size = kernel_size
        self.strides = strides
        self.padding = padding
        self.dilation_rates = dilation_rates
        self.activation = activation
        self.use_bias = use_bias
        super().__init__(*args, **kwargs)

    def build(self, input_shape):
        self.conv = Conv2D(
            self.filters,
            self.kernel_size,
            padding=self.padding,
            dilation_rate=self.dilation_rates[0]
        )
        self.act1 = Activation(self.activation)
        self.act2 = Activation(self.activation)

    def call(self, inputs, **kwargs):
        x1 = self.conv(inputs)
        x2 = tf.nn.conv2d(
                    inputs,
                    self.conv.weights[0], 
                    padding=self.padding,
                    strides=self.strides,
                    dilations=self.dilation_rates[1]
               )
         if self.use_bias:
             x2 = x2 + self.conv.weights[1]
        return x1, x2

Of course, you can also assign a variable and bias directly using `self.add_weights` inside `build`,
and only use `tf.nn.conv2d` in `call`

**Comment by amahendrakar** [comment]
@HuaYZhao,
Any updates regarding this, is this still an issue? Thanks!

**Comment by pandrey-fr** [comment]
@amahendrakar As established per the previous discussion thread, both @Susmit-A and I were able to provide with _ad hoc_ custom layers that can be used in this specific context, however the largest question, _i.e._ "how to share weights (in a simple manner) between layers", is still open.

**Comment by HuaYZhao** [comment]
@amahendrakar Okay, there's no more updates to this question right now! Thank you

**Comment by tanzhenyu** [comment]
What do you mean by sharing weights?
Do you intend to:
1) initialize weights of cnn_2 from weights of cnn_1, or
2) let both cnn_1 and cnn_2 use the same set of weights during both forward and backward path? How do you want to update the weights from gradients?

**Comment by pandrey-fr** [comment]
@tanzhenyu I believe the aim is option 2: letting both layers use the same set of weights (otherwise the framework already allows that, _e.g._ using `get_weights` / `set_weights`), as is done in some modern architectures (_e.g._ ALBERT with transformer layers). I suppose weights update would rely on averaging the gradients computed at each layer's level? From my experience, this is already workable when using similarly-specified layers, simply by iteratively calling a single layer instance, but the question here is to do so with layers that have slightly different specifications (but same number of weights).

**Comment by tanzhenyu** [comment]
> @tanzhenyu I believe the aim is option 2: letting both layers use the same set of weights (otherwise the framework already allows that, _e.g._ using `get_weights` / `set_weights`), as is done in some modern architectures (_e.g._ ALBERT with transformer layers). I suppose weights update would rely on averaging the gradients computed at each layer's level? From my experience, this is already workable when using similarly-specified layers, simply by iteratively calling a single layer instance, but the question here is to do so with layers that have slightly different specifications (but same number of weights).

I see. In that case, what we would recommend is the same way you would share any other python objects, and yes 1) either re-calling a single layer instance, or 2) create a customized layer that does both things, for example [SubwordEncoder](https://www.tensorflow.org/datasets/api_docs/python/tfds/features/text/SubwordTextEncoder) would use the table for both encode and decode.

**Comment by tanzhenyu** [comment]
Closing it. Let us know if you have further questions!

**Comment by lminer** [comment]
@Susmit-A How would you deal with `get_config` in your example with respect to the `filters` argument?
